reV/bespoke/gradient_free.py

```diff
@@ -0,0 +1,247 @@
+        self.offspring_population = np.zeros_like(self.parent_population)
+
+    def initialize_fitness(self):
+        """initialize the fitness of member of the parent population"""
+        # initialize the fitness arrays
+        self.parent_fitness = np.zeros(self.population_size)
+        self.offspring_fitness = np.zeros(self.population_size)
+
+        # initialize fitness of the parent population
+        for i in range(self.population_size):
+            self.chromosome_2_variables(self.parent_population[i])
+            self.parent_fitness[i] = \
+                self.objective_function(self.design_variables)
+
+    def chromosome_2_variables(self, chromosome):
+        """convert the binary chromosomes to design variable values"""
+
+        first_bit = 0
+        float_ind = 0
+
+        for i in range(self.nvars):
+            binary_value = 0
+            for j in range(self.bits[i]):
+                binary_value += chromosome[first_bit + j] * 2**j
+            first_bit += self.bits[i]
+
+            if self.variable_type[i] == "float":
+                self.design_variables[i] = \
+                    self.discretized_variables["float_var%s"
+                                               % float_ind][binary_value]
+                float_ind += 1
+
+            elif self.variable_type[i] == "int":
+                self.design_variables[i] = self.bounds[i][0] + binary_value
+
+    def crossover(self):
+        """perform crossover between individual parents"""
+        self.offspring_population[:, :] = self.parent_population[:, :]
+
+        # mate conscutive pairs of parents (0, 1), (2, 3), ...
+        # The population is shuffled so this does not need to be randomized
+        for i in range(int(self.population_size / 2)):
+            # trade bits in the offspring
+            crossover_arr = np.random.rand(self.nbits)
+            for j in range(self.nbits):
+                if crossover_arr[j] < self.crossover_rate:
+                    self.offspring_population[2 * i][j], \
+                        self.offspring_population[2 * i + 1][j] = \
+                        self.offspring_population[2 * i + 1][j], \
+                        self.offspring_population[2 * i][j]
+
+    def mutate(self):
+        """randomly mutate bits of each chromosome"""
+        for i in range(int(self.population_size)):
+            # mutate bits in the offspring
+            mutate_arr = np.random.rand(self.nbits)
+            for j in range(self.nbits):
+                if mutate_arr[j] < self.mutation_rate:
+                    self.offspring_population[i][j] = \
+                        (self.offspring_population[i][j] + 1) % 2
+
+    def optimize_ga(self):
+        """run the genetic algorithm"""
+
+        self.initialize_design_variables()
+        self.initialize_bits()
+        if self.population_size % 2 == 1:
+            self.population_size += 1
+        self.initialize_population()
+        self.initialize_fitness()
```

Comment:
any reason these init methods shouldnt be called in the `__init__`? Is `optimize_ga()` intended to be called frequently? I wouldnt think so unless you were just running a bunch of random trials, but in that case i would think initializing many GA objects would be the intended pattern.

Reply:
Agreed. I've moved these init methods to __init__